app/models/ancestor_ref.rb

```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class AncestorRef < ActiveRecord::Base
+  DEFAULT_ANCESTOR_BRANCH = 'master'
+
+  fields do
+    ref          :string, limit: 40,  required: true, default: 'master'
+    service_name :string, limit: 255, required: true
+  end
+
+  validates_presence_of :service_name, :ref
+  validates_uniqueness_of :service_name
```

Comment:
These are good to help readability of error messages, and I wanted to add one thing to consider.

In general it's best to also also add an `index` on the `service_name` indicating it's unique so that the DB also has that uniqueness check. This is so we could 100% rely on uniqueness since changes at the DB level should be atomic, meaning the action is indivisible (i.e. nothing else can happen in the middle of this action, everything else must happen before or after). This ensures that multiple changes made to the DB couldn't fall into a common "uniqueness" trap:
```
Change 1a -> check DB to see if service_name "web" exists in DB, sees that there are none
Change 1b -> check DB to see iv service_name "web" exists in DB, sees that there are none
Change 1a -> creates record with service_name: "web"
Change 1b -> creates record with service_name: "web"
```

With all of that said, it likely will not be necessary in this scenario because we won't likely have too many requests happening at once and racing each other, since this is an internal invoca tool. So I'll leave it up to you if you want to add an index, but I wouldn't consider it necessary in this case.

Reply:
ahh that makes sense, thanks for being so thorough on your comment! As I did the research for this work, I saw other models had that `index [:column_name], unique: true`, but I couldn't quite follow how that worked, your explanation helped!

I added the index on service name to be explicit